src/Http/Authentication.Core/src/AuthenticationSchemeProvider.cs

```diff
@@ -52,9 +52,18 @@ protected AuthenticationSchemeProvider(IOptions<AuthenticationOptions> options,
     private IEnumerable<AuthenticationScheme> _requestHandlersCopy = Array.Empty<AuthenticationScheme>();
 
     private Task<AuthenticationScheme?> GetDefaultSchemeAsync()
-        => _options.DefaultScheme != null
-        ? GetSchemeAsync(_options.DefaultScheme)
-        : Task.FromResult<AuthenticationScheme?>(null);
+    {
+        if (_options.DefaultScheme != null)
+        {
+            return GetSchemeAsync(_options.DefaultScheme);
+        }
+        // If no default scheme is specified, and we only have a single scheme, use that
+        if (!_options.DisableAutoDefaultScheme && _schemes.Count == 1)
+        {
+            return GetSchemeAsync(_schemes.First().Key);
+        }
+        return Task.FromResult<AuthenticationScheme?>(null);
```

Comment:
Why do it this late?

Reply:
manipulating the options directly has visible sideeffects to the world.  If we do it this way, we can ensure that this behavior change only affects the execution, it doesn't change the options at all

Reply:
Basically this limits the change to the runtime behavior, as opposed to trying to pick a point in time which we set the default scheme, and then they could turn it off/null it out again after

Reply:
Or to explain this another way, the logic is real consumer of the DefaultScheme setting, so we can guarantee we are doing the right thing in the main codepaths.  The problem with doing it on the options side, is you can add a scheme after and then we are broken.  Lets say we added a PostConfigure to apply AutoDefaultScheme, then the app adds their own `PostConfigure(o => o.AddScheme("YoloASuperLateScheme"))`, we'd have applied the default scheme when there are two schemes.  Doing it this way, we at least guarantee options are consistent, and in addition, we also correctly handle situations where apps dynamically add schemes, since authN schemes are not static for the app.

Reply:
Can we do it in a way where the check only happens once and not per call then?

Reply:
See https://github.com/dotnet/aspnetcore/blob/main/src/Security/samples/DynamicSchemes/Controllers/AuthController.cs#L35 for the sample we have where you can add schemes dynamically

Reply:
Sure, so cache it, and invalidate it if any schemes are added/removed

Reply:
Kinda nasty, but I think I did it correctly within the locks, and added a gross test that adds/removes a bunch of schemes for correctness